src/libraries/System.Net.ServerSentEvents/src/System/Net/ServerSentEvents/SseFormatter.cs

```diff
@@ -0,0 +1,161 @@
+        {
+            using PooledByteBufferWriter bufferWriter = new();
+            using PooledByteBufferWriter userDataBufferWriter = new();
+
+            await foreach (SseItem<T> item in source.WithCancellation(cancellationToken).ConfigureAwait(false))
+            {
+                itemFormatter(item, userDataBufferWriter);
+
+                FormatSseEvent(
+                    bufferWriter,
+                    eventType: item._eventType, // Do not use the public property since it normalizes to "message" if null
+                    data: userDataBufferWriter.WrittenMemory.Span,
+                    eventId: item.EventId,
+                    reconnectionInterval: item.ReconnectionInterval);
+
+                await destination.WriteAsync(bufferWriter.WrittenMemory, cancellationToken).ConfigureAwait(false);
+
+                userDataBufferWriter.Reset();
+                bufferWriter.Reset();
+            }
+        }
+
+        private static void FormatSseEvent(
+            PooledByteBufferWriter bufferWriter,
+            string? eventType,
+            ReadOnlySpan<byte> data,
+            string? eventId,
+            TimeSpan? reconnectionInterval)
+        {
+            Debug.Assert(bufferWriter.WrittenCount is 0);
+
+            if (eventType is not null)
+            {
+                Debug.Assert(!eventType.Contains('\n'));
+
+                bufferWriter.WriteUtf8String("event: "u8);
+                bufferWriter.WriteUtf8String(eventType);
+                bufferWriter.WriteUtf8String(s_newLine);
+            }
+
+            WriteLinesWithPrefix(bufferWriter, prefix: "data: "u8, data);
+            bufferWriter.Write(s_newLine);
+
+            if (eventId is not null)
+            {
+                Debug.Assert(!eventId.Contains('\n'));
+
+                bufferWriter.WriteUtf8String("id: "u8);
+                bufferWriter.WriteUtf8String(eventId);
+                bufferWriter.WriteUtf8String(s_newLine);
+            }
+
+            if (reconnectionInterval is { } retry)
+            {
+                Debug.Assert(retry >= TimeSpan.Zero);
+
+                bufferWriter.WriteUtf8String("retry: "u8);
+                bufferWriter.WriteUtf8Number((long)retry.TotalMilliseconds);
+                bufferWriter.WriteUtf8String(s_newLine);
+            }
+
+            bufferWriter.WriteUtf8String(s_newLine);
+        }
+
+        private static void WriteLinesWithPrefix(PooledByteBufferWriter writer, ReadOnlySpan<byte> prefix, ReadOnlySpan<byte> data)
+        {
+            // Writes a potentially multi-line string, prefixing each line with the given prefix.
+            // Both \n and \r\n sequences are normalized to \n.
+
+            while (true)
+            {
+                writer.WriteUtf8String(prefix);
+
+                int i = data.IndexOf((byte)'\n');
```

Comment:
The spec implies that \r, \n, or \r\n should be used as line separators, but this is only looking for \n and \r\n.
```
Lines must be separated by either a U+000D CARRIAGE RETURN U+000A LINE FEED (CRLF) character pair, a single U+000A LINE FEED (LF) character, or a single U+000D CARRIAGE RETURN (CR) character.
```

Reply:
Fixed.